databend-jdbc/src/main/java/com/databend/jdbc/DatabendConnection.java

```diff
@@ -635,16 +675,35 @@ DatabendClient startQueryWithFailover(String sql, StageAttachment attach) throws
                 throw new SQLException("Error start query: " + "SQL: " + sql + " " + e.getMessage() + " cause: " + e.getCause(), e);
             }
             try {
+                // route hint is used when transaction occured,
+                // transaction procedure:
+                // 1. server return session body where txn state is active
+                // 2. when there is a active transaction, it will route all query to target route hint uri if exists
+                // 3. if there is not active transaction, it will use load balancing policy to choose a host to execute query
+                String query_id = UUID.randomUUID().toString();
+                String candidateHost = this.driverUri.getUri(query_id).toString();
                 if (!inActiveTransaction()) {
-                    this.routeHint = randRouteHint();
+                    this.routeHint = uriRouteHint(candidateHost);
                 }
+                // checkout the host to use from route hint
+                if (this.routeHint != null && !this.routeHint.isEmpty()) {
+                    URI uri = parseRouteHint(this.routeHint);
+                    if (uri != null) {
+                        candidateHost = uri.toString();
+                    }
+                }
+
                 // configure query and choose host based on load balancing policy.
-                ClientSettings.Builder sb = this.makeClientSettings();
+                ClientSettings.Builder sb = this.makeClientSettings(query_id, candidateHost);
                 if (attach != null) {
                     sb.setStageAttachment(attach);
                 }
                 ClientSettings s = sb.build();
                 logger.log(Level.FINE, "retry " + i + " times to execute query: " + sql + " on " + s.getHost());
+                // discover new hosts in need.
+                if (this.autoDiscovery) {
+
```

Comment:
What the empty line for?

Reply:
it should be removed, it is for auto discovery not for this PR.